fw/http_parser.c

```diff
@@ -9039,6 +9039,63 @@ tfw_h2_parse_req_finish(TfwHttpReq *req)
 	return T_OK;
 }
 
+void
+tfw_idx_hdr_parse_host_port(TfwHttpReq *req, TfwStr *hdr)
+{
+	TfwStr *c, *end;
+	bool has_exp_port = false;
+
+	TFW_STR_FOR_EACH_CHUNK(c, hdr, end) {
+		if (!(c->flags & TFW_STR_VALUE) && c->len == 1 &&
+			c->data[0] == ':') {
+			has_exp_port = true;
+			break;
+		}
+	}
+
+	if (has_exp_port) {
+		/* chunk containing the port should always be the last one */
+		TfwStr p_chunk = hdr->chunks[hdr->nchunks - 1];
+		unsigned long host_port = 0;
+		__parse_ulong_ws(p_chunk.data, p_chunk.len, &host_port, USHRT_MAX);
+		T_DBG3("%s: got port: %lu\n", __func__, host_port);
+		req->host_port = host_port;
+	}
+}
+
+void
+tfw_idx_hdr_parse_if_mod_since(TfwHttpReq *req, TfwStr *hdr)
+{
+	TfwStr *c, *end;
+
+	TFW_STR_FOR_EACH_CHUNK(c, hdr, end) {
+		if (c->flags & TFW_STR_HDR_VALUE) {
+			int ret = __h2_req_parse_if_msince((TfwHttpMsg *)req,
+					c->data, c->len, true);
+			T_DBG3("%s: __h2_req_parse_if_msince ret=%d\n",
+				__func__, ret);
+		}
+	}
+
+	T_DBG3("%s: req->cond.m_date: %lu\n", __func__, req->cond.m_date);
+}
+
+void
+tfw_idx_hdr_parse_method(TfwHttpReq *req, TfwStr *hdr)
+{
+	TfwStr *c, *end;
+
+	TFW_STR_FOR_EACH_CHUNK(c, hdr, end) {
+		if (c->flags & TFW_STR_HDR_VALUE) {
+			int ret = tfw_h2_parse_req_hdr(c->data, c->len, req,
+							true, true);
```

Comment:
The parser function is heavy and it's not good to call the whole parser function just for method. If we 100% need to call the parser from hpack, then we probably need to move the method parsing into a separate small function. However, it seems this is a misconception since `tfw_hpack_hdr_set()` is called from `HPACK_STATE_ALL_INDEXED`, which implies that the whole header is fully indexed.
 
The current approach is `tfw_h2_parse_req()` -> `tfw_hpack_decode()` (decode indexes and Huffman) -> `HPACK_DECODE_PROCESS_STRING()`, which leads to `__hpack_process_hdr_value()` or `__hpack_process_hdr_name()` for Huffman encoded strings, or `HPACK_PROCESS_STRING(), which leads to `tfw_h2_parse_req_hdr()` for plain strings.

It seems the bug is in the state machine inside `tfw_hpack_decode()` and we should still process `Host` and `If-Modifies-Since` for the test as `HPACK_STATE_VALUE_TEXT`.